src/motion/features/viewport/observers.ts

```diff
@@ -0,0 +1,71 @@
+type IntersectionHandler = (entry: IntersectionObserverEntry) => void
+
+interface ElementIntersectionObservers {
+    [key: string]: IntersectionObserver
+}
+
+/**
+ * Map an IntersectionHandler callback to an element. We only ever make one handler for one
+ * element, so even though these handlers might all be triggered by different
+ * observers, we can keep them in the same map.
+ */
+const observerCallbacks = new WeakMap<Element, IntersectionHandler>()
+
+/**
+ * Multiple observers can be created for multiple element/document roots. Each with
+ * different settings. So here we store dictionaries of observers to each root,
+ * using serialised settings (threshold/margin) as lookup keys.
+ */
+const observers = new WeakMap<
+    Element | Document,
+    ElementIntersectionObservers
+>()
+
+const fireObserverCallback = (entry: IntersectionObserverEntry) => {
+    observerCallbacks.get(entry.target)?.(entry)
+}
+
+const fireAllObserverCallbacks: IntersectionObserverCallback = (entries) => {
+    entries.forEach(fireObserverCallback)
+}
+
+function initIntersectionObserver({
+    root,
+    ...options
+}: IntersectionObserverInit): IntersectionObserver {
+    /**
+     * If we don't have an observer lookup map for this root, create one.
+     */
+    if (!observers.has(root || document)) {
+        observers.set(root || document, {})
+    }
+    const rootObservers = observers.get(root || document)!
```

Comment:
Why don't we assign `root || document` to a constant here?
And shouldn't it be `root ?? document`?

Reply:
And this should be called `intersectionObservers`, right?

Reply:
I don't see why it'd be `??` but I can assign to a var

Reply:
The name was chosen as it's specifically the subset of intersection observers we're storing for this particular root

Reply:
Ah, I somehow completely missed that we're using a object keyed by root here 🤦🏻